### PR TITLE
types: create jsdoc based typescript typings

### PR DESCRIPTION
**Pull request by ChristianMurphy**

### CODE REVIEW

File: lib/evaluate.js [hunk 1]
```diff
@@ -1,20 +1,40 @@
...
+ * @param {VFileCompatible} file 
+ * @param {*} options 
+ */
 export async function evaluate(file, options) {
   var config = split(options)
   // What is this?
   // V8 on Node 12 can’t handle `eval` for collecting coverage, apparently…
   /* c8 ignore next 2 */
-  return new Function(await compile(file, config.compile))(config.run)
+  return new Function((await compile(file, config.compile)).toString())(config.run)
```
Comment (ChristianMurphy) [inline lib/evaluate.js:20]: explicit cast from vfile to string for typescript

File: test/core.js [hunk 2]
```diff
@@ -844,19 +844,19 @@ test('MDX (ESM)', async function (t) {
   )
 
   t.equal(
-    await run(compileSync('export const number = Math.pi'), {
+    await run(compileSync('export const number = Math.PI'), {
       returnModule: true
     }).number,
-    Math.pi,
+    Math.PI,
```
Comment (ChristianMurphy) [inline test/core.js:850]: `Math.pi` is undefined, the constant `Math.PI` is

File: lib/evaluate.js [hunk 3]
```diff
@@ -1,20 +1,57 @@
...
+/**
+ * @typedef {import("vfile").VFileCompatible} VFileCompatible
+ * @typedef {import("unified").Pluggable} Pluggable
+ * @typedef {import('./core.js').ProcessorOptions} ProcessorOptions
+ *
+ * @typedef RunnerOptions
+ * @property {*} [Fragment]
+ * @property {*} [jsxs]
+ * @property {*} [recastPlugins]
+ * @property {*} [useMDXComponents]
```
Comment (ChristianMurphy) [inline lib/evaluate.js:14]: more definition around what these could/should be could be helpfup

File: lib/recma-stringify.js [hunk 4]
```diff
@@ -1,12 +1,31 @@
...
 import {baseGenerator, generate} from 'astring'
 import stringifyEntities from 'stringify-entities/light.js'
 
-// A small wrapper around `astring` to add support for serializing JSX.
+/**
+ * @typedef {import("unist").Node} Node
+ * @typedef {import("vfile").VFile} VFile
+ *
+ * @typedef RecmaStringifyOptions
+ * @property {*} [SourceMapGenerator]
```
Comment (ChristianMurphy) [inline lib/recma-stringify.js:10]: does this specifically have to be the source map generator library?
Or are there others that could be used?

File: lib/remark-mdx.js [hunk 5]
```diff
@@ -3,7 +3,12 @@ import {fromMarkdown, toMarkdown} from 'mdast-util-mdx'
 
 var warningIssued
 
-// Add the micromark and mdast extensions for MDX.js (JS aware MDX).
+/**
+ * Add the micromark and mdast extensions for MDX.js (JS aware MDX).
+ *
+ * @param {*} [options]
```
Comment (ChristianMurphy) [inline lib/remark-mdx.js:9]: `micromark-extension-mdxjs` would need typings to complete this

File: lib/webpack-loader.js [hunk 6]
```diff
@@ -1,10 +1,14 @@
...
-// See `webpack.cjs`, which wraps this, because Webpack loaders must currently
-// be CommonJS.
-// `file.map` is defined when a `SourceMapGenerator` is passed in options.
+/**
+ * A Webpack (4+) loader for xdm.
+ * See `webpack.cjs`, which wraps this, because Webpack loaders must currently
+ * be CommonJS.
+ * `file.map` is defined when a `SourceMapGenerator` is passed in options.
+ * @param {string} value
+ * @param {*} callback
```
Comment (ChristianMurphy) [inline lib/webpack-loader.js:10]: haven't found the exact callback construct in the webpack docs or typings yet

Comment (wooorm) [reply lib/webpack-loader.js:10]: `callback` comes from `this.async()` https://github.com/wooorm/xdm/blob/21c18aeea2e0f29cdbac7488ef54745ff8f749e8/webpack.cjs#L5

File: lib/recma-stringify.js (hunk 4 above)
Comment (wooorm) [reply lib/recma-stringify.js:10]: yes, specifically from `source-map`

File: test/rollup.js [hunk 7]
```diff
@@ -19,10 +19,11 @@ test('xdm (rollup)', async function (t) {
     external: ['react/jsx-runtime'],
     plugins: [
       {
+        name: 'xdm',
         async transform(contents, filePath) {
           if (path.extname(filePath) !== '.mdx') return null
           var file = await compile({contents, path: filePath})
-          return {code: file.contents, map: file.map}
+          return {code: String(file.contents), map: file.map?.toString()}
```
Comment (wooorm) [inline test/rollup.js:26]: `toString` is incorrect: `file.map` is an (optional) JSON object.

Also: passing the JSON should be fine: https://github.com/rollup/rollup/blob/850d30a463b0b50f409413081629e33c81139300/src/rollup/types.d.ts#L101

(btw: rollup can apparently also get an acorn AST! that might be a nice performance optimization! I looked into passing ASTs for webpack’s babel loader but it doesn’t support it apparently)

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        "index.js"
+      ],
+      "include": [
+        "lib/**/*.js",
+        "test/**/*.js"
+      ],
+      "compilerOptions": {
+        "target": "ES2020",
+        "lib": ["ES2020"],
+        "module": "commonjs",
```
Comment (wooorm) [inline tsconfig.json:12]: is `commonjs` (even as a compiler option) correct?
If these lines are copied from another place, this is a good time to check if some can be omitted

File: lib/recma-jsx-rewrite.js [hunk 9]
```diff
@@ -3,14 +3,30 @@ import {walk} from 'estree-walker'
 import {analyze} from 'periscopic'
 import u from 'unist-builder'
 
-// A plugin that rewrites JSX in functions to accept components as
-// `props.components` (when the function is called `MDXContent`), or from
-// a provider (if there is one).
-// It also makes sure that any undefined components are defined: either from
-// received components or as a function that throws an error.
+/**
+ * @typedef {import("periscopic").Node} Node
```
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply lib/recma-jsx-rewrite.js:7]: I tried using using `import("estree").Node` the issue comes in on line 32.
`analyze` uses the definition from line 60 https://unpkg.com/browse/periscopic@3.0.0/types/index.d.ts
Which is not compatible with the `import("estree").Node` that can be directly imported, either due to an override or due to a version conflict.

[the remark this answers is not in the document]
Comment (wooorm) [reply lib/recma-jsx-rewrite.js:7]: The results from the use of periscopic here is indeed only an identifier node (say, a strong node for mdast).
But the rest of the tree is a complete estree node.

Anyway, I don’t know enough about TS to have a meaningful idea here, but I’d guess that using the whole estree types here should be possible, and “narrowing” it to identifiers for the result of periscopic could also work?

[the remark this answers is not in the document]
Comment (wooorm) [reply lib/recma-jsx-rewrite.js:7]: ohh, I guess their Node is too specific compared to an estree Node? 🤔 Weird.

[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply lib/recma-jsx-rewrite.js:7]: I'm a bit confused by it, https://unpkg.com/browse/periscopic@3.0.0/types/index.d.ts line 1 indicates it should be the same as `estree`, but line 60 appears to make it more specific/narrow... :thinking:

File: test/core.js [hunk 10]
```diff
@@ -106,7 +106,9 @@ test('xdm', async function (t) {
   )
 
   t.equal(
-    render(h(await run(compileSync('?', {jsxImportSource: 'preact/compat'})))),
+    render(
+      h(await run(compileSync('?', {jsxImportSource: 'preact/compat'})), {})
```
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply test/core.js:110]: according to
https://github.com/preactjs/preact/blob/9250ac91891d19ec988938971b6767697bcb028e/src/index.d.ts#L184-L215
yep, there is no version of `createElement` or `h` that has `props` as being optional

File: lib/evaluate.js [hunk 11]
```diff
@@ -1,20 +1,57 @@
...
+ *
+ * @typedef {ProcessorOptions & RunnerOptions} ProcessorAndRunnerOptions
+ * @typedef {{ [name: string]: any }} ComponentMap
+ * @typedef {{ [props: string]: any, components?: ComponentMap }} XDMProps
```
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply lib/evaluate.js:18]: applied the first, can you expand what you mean by the second?

File: test/rollup.js [hunk 12]
```diff
@@ -20,10 +20,11 @@ test('xdm (rollup)', async function (t) {
     plugins: [
       {
         name: 'xdm',
+        // @ts-ignore TODO narrow the types of file.contents and file.map
         async transform(contents, filePath) {
           if (path.extname(filePath) !== '.mdx') return null
           var file = await compile({contents, path: filePath})
-          return {code: String(file.contents), map: file.map?.toString()}
+          return {code: file.contents, map: file.map}
```
Comment (wooorm) [inline test/rollup.js]: here’s an example that shows it works: https://github.com/rollup/plugins/blob/f4999af9ebcc495a0a0f43aeba0c3cc820a58ee6/packages/dsv/src/index.js#L29

Comment (ChristianMurphy) [reply test/rollup.js]: right, the issue here is that `VFile` has `contents` as being `string | Uint8Array` and `map` being `unknown` since it is not a standard `VFile` property.
To make this seemless in typescript both of these would need to be narrowed to be compatible with the rollup type.

File: tsconfig.json (hunk 8 above)
Comment (ChristianMurphy) [reply tsconfig.json]: it's a quick way to set `moduleResolution` to `node`, updated to use `module: ES2020` with an explicit `moduleResolution`

File: test/rollup.js (hunk 12 above)
Comment (wooorm) [reply test/rollup.js]: Oh, I’m fine with `String(file.contents)` btw, I think my problem was with `map: file.map.toString()` (because that would yield `'undefined'` or `[object Object]`)

File: lib/webpack-loader.js (hunk 6 above)
Comment (ChristianMurphy) [reply lib/webpack-loader.js]: thanks! updated

File: test/rollup.js (hunk 12 above)
Comment (ChristianMurphy) [reply test/rollup.js]: :man_shrugging: if you prefer the cast I can add it back, the type conflict with `map` would remain, so either way there will be a `@ts-ignore` until this can be refined.

Comment (wooorm) [reply test/rollup.js]: I have no preference for `file.contents`, whatever works for Rollup/TS!

**Comment by wooorm** [comment]
One extra thing is that TS is probably not reading `webpack.cjs` in the root, due to its extension? That might be useful for webpack folks wanting types

File: test/rollup.js (hunk 12 above)
Comment (ChristianMurphy) [reply test/rollup.js]: keeping as it originally was for now, may revisit down the road.

**Comment by ChristianMurphy** [comment]
I can add jsdocs to `webpack.cjs`, but the extension will conflict with

```
error TS6054: File 'xdm/webpack.cjs' has an unsupported extension. The only supported extensions are '.ts', '.tsx', '.d.ts', '.js', '.jsx'.
```

adding cjs extension support in typescript is an open item at https://github.com/microsoft/TypeScript/issues/38784

File: test/core.js [hunk 13]
```diff
@@ -279,6 +283,7 @@ test('xdm', async function (t) {
         {
           components: {
             y: {
+              // @ts-ignore React typings do not support nested component names
```
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply test/core.js:286]: Tracing it back more, it looks like this found an edge case not covered in https://github.com/mdx-js/mdx/blob/39b6f23bef8401b15d0f2f66031647b703952fb3/packages/react/types/index.d.ts#L15-L26

File: lib/evaluate.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (wooorm) [reply lib/evaluate.js]: The result of `evaluate()` at `.default` is a component: a function that takes `MDXContentProps` and returns a call to `jsx`: in React, I’m sure there’s a type for that.

Now, assuming `jsx` had types in React, then there’d be a type for its return value, right? Then I’m guessing that you could take that value and define it as `.default`s return value too?

Illustrated in another way: `jsx` is a function: `f() = x`. So the return value of `.default` is also always `x`

(and for my comment I forgot that `pragma` is not an option for `evaluate`, and the return value of `jsxs` and `jsx` are the same)

File: test/core.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (wooorm) [reply test/core.js:110]: Funky. It should be, because the code allows it, but that’s for them to fix 🤷‍♂️

File: test/core.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (wooorm) [reply test/core.js:286]: Ah. `mdx-js/mdx` does not allow `<a.b />`, this was fixed there and is also supported here`.
Why does that affect this test though?

File: lib/recma-jsx-rewrite.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (wooorm) [reply lib/recma-jsx-rewrite.js:7]: And would actually using `estree` for the other places (https://github.com/wooorm/xdm/pull/2/files/d347c39c923aa71cfd19e6801177f0f22d0e6877#diff-ca1bb3ad6385be868a6efa965db2bbb1da4fedd201ae4fc1f11e3d8923d95115R41-R48) help?

File: lib/evaluate.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply lib/evaluate.js]: > The result of evaluate() at .default is a component: a function that takes MDXContentProps

Right, and that is the `ExportMap` type on line 19

> Now, assuming jsx had types in React, then there’d be a type for its return value, right?

For react yes,
one that is consistent across React, Preact, and Vue, I'm not so sure.
I think there are differences (it would take some experimentation to check), and if there are differences, it would be difficult to capture even with conditional types watching what was passed to `pragma`.

[the remark this answers is not in the document]
Comment (wooorm) [reply lib/evaluate.js]: There would be difference but the result of that `f() = x` should be accessible though?

If someone does `evaluateSync({jsx: f}).default()` the output type is still `x`, and `jsx` is a required option.

I understand that all the things passed as `jsx` will be different, but they’re required to be functions, so that might help?

[the remark this answers is not in the document]
Comment (wooorm) [reply lib/evaluate.js]: Preact’s runtime does have types: https://github.com/preactjs/preact/blob/9250ac91891d19ec988938971b6767697bcb028e/jsx-runtime/src/index.d.ts#L11-L17

File: lib/recma-jsx-rewrite.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply lib/recma-jsx-rewrite.js:7]: It may not be a bad idea to use `estree.Node` in other places.
It wouldn't fix this particular issue, which is a conflict between the signature of the method and a function call inside the method.

File: test/core.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply test/core.js:286]: https://github.com/wooorm/xdm/blob/d347c39c923aa71cfd19e6801177f0f22d0e6877/test/core.js#L257

is generic, it will try to match the props defined by

https://github.com/wooorm/xdm/blob/d347c39c923aa71cfd19e6801177f0f22d0e6877/test/core.js#L258

with the props passed here

https://github.com/wooorm/xdm/blob/d347c39c923aa71cfd19e6801177f0f22d0e6877/test/core.js#L259-L265

[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply test/core.js:286]: which means the props here, would need to match https://github.com/mdx-js/mdx/blob/39b6f23bef8401b15d0f2f66031647b703952fb3/packages/react/types/index.d.ts#L15-L26
or a `ts-ignore` would need to be applied

[the remark this answers is not in the document]
Comment (wooorm) [reply test/core.js:286]: Ah... OK, that explains it. Well, then this is fine for me!

File: lib/recma-jsx-rewrite.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply lib/recma-jsx-rewrite.js]: smoothed over this with an ignore for consistency of using estree (b466ab6)

File: lib/recma-jsx-build.js [hunk 14]
```diff
@@ -41,6 +37,7 @@ export function recmaJsxBuild({contain}) {
                 u('Property', {
                   kind: 'init',
                   shorthand: false,
+                  // @ts-ignore assume specifier is not a default export
```
Comment (ChristianMurphy) [inline lib/recma-jsx-build.js:40]: This is something to note, I'm not sure if it's a quirk of the typing or potentially an edge case not covered

Comment (wooorm) [reply lib/recma-jsx-build.js:40]: `build` does not generate a default export, to in this case it’s fine!

File: test/babel.js [hunk 15]
```diff
@@ -39,6 +41,7 @@ test('xdm (babel)', async function (t) {
         ) {
           return compileSync(
             {contents, path: options.sourceFileName},
+            // @ts-ignore TODO find out why compiler causes TS error
             {recmaPlugins: [recmaBabel]}
           ).result
```
[the remark this answers is not in the document]
Comment (ChristianMurphy) [reply test/babel.js:46]: That may be the case

**Comment by wooorm** [comment]
`find . -maxdepth 3 -type f -name "*.d.ts"` was removing a ton of stuff from `node_modules` btw!
